### ISSUE TITLE
Unreachable code error not reported when type is never

### ISSUE THREAD

**Issue by u130b8** [microsoft/TypeScript#60212]
### 🔎 Search Terms

unreachable code, never

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about this

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/MYGwhgzhAECqEFMBO0DeAoaXoEsAmAXNAHYCuAtgEbLQC80ADANzoC+66AZqccAC44A9sWicAFKURIi8ZAEo0mbDk7QJUgHT5oAQlr0yIEAozYz0JAj6kkxFmfZKsk5FrxNoT6AHpv2AHoA-GxAA

### 💻 Code

```ts
class User {
    id: number = 0;
}

function f(user: User) {
    if (user.id !== null) {
        return;
    }

    user.id; // never
    //   ^?
}
```

### 🙁 Actual behavior

Unreachable code error is not reported

### 🙂 Expected behavior

Unreachable code error should be reported, or a warning that `user.id !== null` is a constant condition as `user.id` is always a `number`. 

### Additional information about the issue

I was refactoring TS code and the `user.id` property used to be `number | null`:

```ts
class User {
    id: number|null = null;
}

function maybeInitialize(user: User) {
    if (user.id !== null) {
        // User already initialized, do nothing
        return;
    }

    // Run important initialization code
    user.id = 123;
}
```
I then changed it to be just `number`, but the old code that assumed `number | null` was still there, even though it is now unreachable:

```ts
class User {
    id: number = 0;
}

function maybeInitialize(user: User) {
    if (user.id !== null) {
        // This is now effectively `true`, and always runs
        return;
    }
    
    // Important code that will never run anymore, no error from TS
    user.id = 123;
}
```

TS does detect that it's `never` after the check but doesn't report it as unreachable code even with `"allowUnreachableCode": false`.

I think this should definitely be reported as unreachable code error as one of the main use cases of TS for me is compiler-assisted refactoring, where I can easily change the type and all of the places that assumed it to be a different type are reported as errors. In this situation, the compiler fails to notify me of this error, even though it is detected as a `never` type.

**Comment by MartinJohns** [comment]
Duplicate of #39995.
